### PR TITLE
I/O statistics counters for Tap/RawSocket apps

### PR DESCRIPTION
**Pull request by eugeneia**
Depends on #931. Coverage of Solarflare and Intel1G apps is coming in a separate PR (waiting for Solarflare test hardware and completion of #897).

Besides adding statistics counters[1] to the Tap and RawSocket apps, this also adds counting of broadcast packets to vhost_user.

1: `provided_counters = {  'type', 'dtime', 'rxbytes', 'rxpackets', 'rxmcast', 'rxbcast', 'txbytes', 'txpackets', 'txmcast',  'txbcast' }`

### CODE REVIEW

File: src/apps/intel/intel10g.lua [hunk 1]
```diff
@@ -452,22 +294,10 @@ end
...
-   --    packet.free(p)
-   -- else
-   do
-      self.txdesc[self.tdt].address = memory.virtual_to_physical(p.data)
-      self.txdesc[self.tdt].options = bor(p.length, txdesc_flags, lshift(p.length+0ULL, 46))
-      self.txpackets[self.tdt] = p
-      self.tdt = band(self.tdt + 1, num_descriptors - 1)
-   end
+   self.txdesc[self.tdt].address = memory.virtual_to_physical(p.data)
+   self.txdesc[self.tdt].options = bor(p.length, txdesc_flags, lshift(p.length+0ULL, 46))
```
Comment (kbara) [inline src/apps/intel/intel10g.lua]: In future patches, I'd like comments or defined constants explaining numbers like that 46.

File: src/apps/intel/intel_app.lua [hunk 2]
```diff
@@ -31,25 +34,53 @@ local function firsthole(t)
...
+      self.stats.path = "/counters/"..conf.pciaddr.."/"
+      self.stats.sync_timer = lib.timer(0.001, 'repeating', engine.now)
+      self.stats.counters = {}
+      for _, name in ipairs(provided_counters) do
+         self.stats.counters[name] = counter.open(self.stats.path..name)
+      end
+      counter.set(self.stats.counters.type, 0x1000) -- Hardware interface
+      counter.set(self.stats.counters.dtime, C.get_unix_time())
+      counter.set(self.stats.counters.mtu, self.dev.mtu)
+      counter.set(self.stats.counters.speed, 10000000) -- 10 Gbits
```
Comment (kbara) [inline src/apps/intel/intel_app.lua]: Why is this a multiple of kilobits?

**Comment by kbara** [comment]
This has merge conflicts with the packet documentation in README.md. Once those are fixed, I'm happy to merge this.

File: src/apps/intel/intel_app.lua (hunk 2 above)
Comment (alexandergall) [inline src/apps/intel/intel_app.lua]: > Why is this a multiple of kilobits?

I guess this has sneaked in from my faulty SNMP code. For SNMP, there are two objects: `ifSpeed` and `ifHighSpeed`. The correct values for 10Gbps are `ifSpeed = 4294967295` and `ifHighSpeed = 10000` (RFC3635). My code used `ifSpeed = 10000000`, which is wrong and made it to here. The `speed` counter has nothing to do with SNMP and should say 10000000000 (hope I counted that right :).

**Comment by eugeneia** [comment]
@kbara merge conflicts with which branch, `kbara-next`?  Current policy would be for upstream (you) to resolve that conflict. If I merged `kbara-next` into this branch this PR would include all commits in `master..kbara-next`

**Comment by kbara** [comment]
It has merge conflicts with kbara-next, due to the documentation patch that you also wrote (on packet structure) and which has been upstreamed. I'll resolve the conflict if you'd prefer not to.

**Pull request merged**